aptos-move/framework/aptos-framework/sources/Comparator.move

```diff
@@ -0,0 +1,146 @@
+/// Provides a framework for comparing two elements
+module AptosFramework::Comparator {
+    use Std::BCS;
+    use Std::Vector;
+
+    const EQUAL: u8 = 0;
+    const SMALLER: u8 = 1;
+    const GREATER: u8 = 2;
+
+    struct Result has drop {
+        inner: u8,
+    }
+
+    public fun is_equal(result: &Result): bool {
+        return result.inner == EQUAL
+    }
+
+    public fun is_smaller_than(result: &Result): bool {
+        return result.inner == SMALLER
+    }
+
+    public fun is_greater_than(result: &Result): bool {
+        result.inner == GREATER
+    }
+
+    // Performs a comparison of two types after BCS serialization.
+    public fun compare<T>(left: &T, right: &T): Result {
+        let left_bytes = BCS::to_bytes(left);
+        let right_bytes = BCS::to_bytes(right);
+
+        compare_u8_vector(left_bytes, right_bytes)
+    }
+
+    // Performs a comparison of two vector<u8>s or byte vectors
+    public fun compare_u8_vector(left: vector<u8>, right: vector<u8>): Result {
+        let left_length = Vector::length(&left);
+        let right_length = Vector::length(&right);
+
+        let idx = 0;
+
+        while (idx < left_length && idx < right_length) {
+            let left_byte = *Vector::borrow(&left, idx);
+            let right_byte = *Vector::borrow(&right, idx);
+
+            if (left_byte < right_byte) {
+                return Result { inner: SMALLER }
+            } else if (left_byte > right_byte) {
+                return Result { inner: GREATER }
+            };
+            idx = idx + 1;
+        };
+
+        if (left_length < right_length) {
+            Result { inner: SMALLER }
+        } else if (left_length > right_length) {
+            Result { inner: GREATER }
+        } else {
+            Result { inner: EQUAL }
+        }
+    }
+
+    #[test]
+    public fun test_strings() {
+        use Std::ASCII;
+
+        let value0 = ASCII::string(b"alpha");
+        let value1 = ASCII::string(b"beta");
+        let value2 = ASCII::string(b"betaa");
+
+        assert!(is_equal(&compare(&value0, &value0)), 0);
+        assert!(is_equal(&compare(&value1, &value1)), 1);
+        assert!(is_smaller_than(&compare(&value1, &value0)), 2);
+        assert!(is_greater_than(&compare(&value0, &value1)), 3);
+        assert!(is_smaller_than(&compare(&value0, &value2)), 4);
+        assert!(is_smaller_than(&compare(&value1, &value2)), 5);
+        assert!(is_greater_than(&compare(&value2, &value0)), 6);
+        assert!(is_greater_than(&compare(&value2, &value1)), 7);
```

Comment:
-_- the ordering is a myth

Reply:
are you arguing that the ordering appears inconsistent?

beta < alpha < betaa

because it is (string length, string contents)
bcs is great, we need traits or interfaces....

Reply:
I mean, I'd assert in this order (or something else that I can follow better):
0 < 1
1 > 0
0 < 2
2 > 0
1 < 2
2 > 1

Reply:
That was my second guess, so I already restructured it ;)